packages/@lwc/template-compiler/src/parser/experimental.ts

```diff
@@ -0,0 +1,157 @@
+/*
+ * Copyright (c) 2018, salesforce.com, inc.
+ * All rights reserved.
+ * SPDX-License-Identifier: MIT
+ * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/MIT
+ */
+import * as parse5 from 'parse5-with-errors';
+
+import {
+    CompilerDiagnostic,
+    generateCompilerDiagnostic,
+    LWCErrorInfo,
+    ParserDiagnostics,
+} from '@lwc/errors';
+
+import { parseHTML, traverseHTML, treeAdapter } from './html';
+
+function filter(node: parse5.AST.Default.Element) {
+    delete node.parentNode;
+    delete node.__location;
+
+    let nodeToFilter = node;
+
+    const content = treeAdapter.getTemplateContent(node);
+    if (content) {
+        nodeToFilter = content as parse5.AST.Default.Element;
+    }
+
+    filterChildNodes(nodeToFilter);
+}
+
+function filterChildNodes(node: parse5.AST.Default.Element) {
+    node.childNodes = node.childNodes.filter(childNode => {
+        return !treeAdapter.isTextNode(childNode) && !treeAdapter.isCommentNode(childNode);
+    });
+}
+
+export function experimentalParse(source: string) {
+    const warnings: CompilerDiagnostic[] = [];
+    const { fragment, errors: parsingErrors } = parseHTML(source);
+
+    if (parsingErrors.length !== 0) {
+        return { warnings: parsingErrors };
+    }
+
+    const templateRoot = getTemplateRoot(fragment);
+    if (!templateRoot) {
+        return { warnings };
+    }
+
+    traverseHTML(templateRoot, {
+        Element: {
+            enter(node: parse5.AST.Default.Element) {
+                filter(node);
+            },
+        },
+    });
+
+    const root = templateRoot as unknown;
+    return root;
+
+    function getTemplateRoot(
+        documentFragment: parse5.AST.Default.DocumentFragment
+    ): parse5.AST.Default.Element | undefined {
+        // Filter all the empty text nodes
+        const validRoots = documentFragment.childNodes.filter(
+            child =>
```

Comment:
we can probably eliminate additional iterations, such as on line 77, if we used a regular loop instead and collected two lists, one for text nodes and one root nodes; then the multi-root validation becomes a sum of both lengths and the root retrieval is simply a zeroth index of roots list.

Reply:
I'm not sure. I think the existing code might be easier to understand.